Parse decimal integers in only one place, not four

fingolfin opened the pull request:
This adds a tiny amount of overhead to the parsing stage, as we now sometimes verify that all characters in the input are digits; but this should be negligible except for extreme cases. If we are truly worried about this, we could return to having two versions of this...

ChrisJefferson commented on src/intfuncs.h:
```diff
@@ -47,7 +47,14 @@ Int HASHKEY_WHOLE_BAG_NC (Obj obj, UInt4 seed);
-Obj IntStringInternal( Obj string );
+// Parse a string containing a decimal integer into a GAP integer
+// object. Returns 'fail' if the input is not a valid decimal.
+// This function ignores leading spaces, and handles signs.
```
The statement "This function ignores leading spaces" doesn't seem to be true, either before or after this PR?

fingolfin replied:
You’re right, will fix

ChrisJefferson commented on src/intrprtr.c:
```diff
@@ -1975,24 +1935,20 @@ void            IntrIntExpr (
-    
-    if ( ret == Fail ) {
-        /* This should never happen */
-        ErrorQuit("Int: Invalid parsing (internal error)", 0, 0);
```
I am a little nervous about dumping this entirely -- while I agree it should never happen, it could also cause a serious byte-code problem if it ever did.

Perhaps a `GAP_ASSERT(ret != Fail)`?

fingolfin replied:
For coding, such an asser already exists in `CodeIntExpr`. For interpretation, I don’t see a problem.

Finally, the removed Error check itself is fairly recent, before that (and in the other code paths), we never made comparable checks. And really, if somebody breaks the parser badly enough to let non-digits reach this point, we’d be in deep trouble and notice anyway.

fingolfin replied:
Specifically, you added that “paranoia” check two years ago, see https://github.com/gap-system/gap/commit/d4e4d740c4a06f92ec8a05e1fa463a02977fbfbe

ChrisJefferson replied:
That would explain why I liked it :) if you are still really it can't occur, then I am happy to leave it out.

fingolfin replied:
Yes, I am sure it can't happen at that point. And even if I am totally wrong, then this will be immediately visible and easy to debug, as `fail` would occur in a place where an integer object is expected (and that would have been triggered by a bug in the GAP parser).

That said, being extra paranoid and adding the two `GAP_ASSERT` doesn't hurt, so I just did that.

fingolfin commented:
I updated this PR now. Also moved the integer parsing to `gmpints.c`. I plan to rewrite `IntStringInternal` in a future PR to be more efficient, by immediately creating a GMP integer of the right size, and performing in-place operations, thus avoiding creating tons of intermediate objects. So parsing integers with thousands of digits will become a bit more efficient.
(Alternatively, we could call out to the relevant GMP interface. We'll see.

fingolfin commented:
I‘ll take care of the test failures when I am back home.
